onnxruntime/core/optimizer/embed_layer_norm_fusion.cc

```diff
@@ -0,0 +1,183 @@
+      continue;
+    }
+    Node& reduce_sum_node = *graph.GetNode(edges[0]->GetNode().Index());
+    if (reduce_sum_node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    // Traceback the SkipLayerNormalization node to find Gather --> SkipLayerNormalization
+    std::vector<graph_utils::EdgeEndToMatch> segment_embedding_path{
+        {0, 0, "Gather", {1}, kOnnxDomain}};
+    edges.clear();
+    if (!graph_utils::FindPath(skip_ln_node, true, segment_embedding_path, edges, logger)) {
+      continue;
+    }
+    Node& segment_gather_node = *graph.GetNode(edges[0]->GetNode().Index());
+    if (segment_gather_node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    // Traceback the SkipLayerNormalization node to find Gather --> Add --> SkipLayerNormalization
+    std::vector<graph_utils::EdgeEndToMatch> word_embedding_path{
+        {0, 1, "Add", {7}, kOnnxDomain},
+        {0, 0, "Gather", {1}, kOnnxDomain}};
+    edges.clear();
+    if (!graph_utils::FindPath(skip_ln_node, true, word_embedding_path, edges, logger)) {
+      continue;
+    }
+    Node& add_node = *graph.GetNode(edges[0]->GetNode().Index());
+    Node& word_gather_node = *graph.GetNode(edges[1]->GetNode().Index());
+    if (add_node.GetOutputEdgesCount() != 1 || word_gather_node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    // Traceback the Add node to find Shape --> Expand --> Gather --> Add
+    std::vector<graph_utils::EdgeEndToMatch> position_embedding_path{
+        {0, 1, "Gather", {1}, kOnnxDomain}};
+        //{0, 1, "Expand", {8}, kOnnxDomain},
+        //{0, 1, "Shape", {1}, kOnnxDomain}};
+    edges.clear();
+    if (!graph_utils::FindPath(add_node, true, position_embedding_path, edges, logger)) {
+      continue;
+    }
+    Node& position_gather_node = *graph.GetNode(edges[0]->GetNode().Index());
+    //Node& expand_node = *graph.GetNode(edges[1]->GetNode().Index());
+    //Node& shape_node = *graph.GetNode(edges[2]->GetNode().Index());
+    if (position_gather_node.GetOutputEdgesCount() != 1) {
+      continue;
+    }
+
+    // Get input "input_ids" from node.
+    NodeArg* input_ids = nullptr;
+    if (!graph_utils::NodeArgIsConstant(graph, *(word_gather_node.MutableInputDefs()[1])) &&
+        !graph_utils::IsGraphInput(graph, word_gather_node.MutableInputDefs()[1])) {
+      continue;
+    }
+    input_ids = word_gather_node.MutableInputDefs()[1];
+
+    // Get input "segment_ids" from node.
+    NodeArg* segment_ids = nullptr;
+    if (!graph_utils::NodeArgIsConstant(graph, *(segment_gather_node.MutableInputDefs()[1])) &&
+        !graph_utils::IsGraphInput(graph, segment_gather_node.MutableInputDefs()[1])) {
+      continue;
+    }
+    segment_ids = segment_gather_node.MutableInputDefs()[1];
+
+    // Get input "mask" from "ReduceSum" node.
+    NodeArg* mask = nullptr;
+    if (!graph_utils::NodeArgIsConstant(graph, *(reduce_sum_node.MutableInputDefs()[0])) &&
```

Comment:
No need to check here. It is because the input of ReduceSum could be the following cases:
(1) Original mask input (if it is int32 tensor)
(2) Cast node output (if mask is int64 tensor). It is because Attention fusion will add a Cast node to convert it to int32.
(3) Other node if it is not standard bert model.